### PR TITLE
mgr/rgw: Add usage trim support on interval

### PR DESCRIPTION
**Pull request by tobias-urdin**
This adds support in the rgw ceph-mgr module to trim the usage log on a regular interval so that one does not need to run radosgw-admin inside a cronjob or manually.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/rgw/module.py [hunk 1]
```diff
@@ -181,6 +191,27 @@ def _cmd_rgw_realm_reconcile(self,
...
+            if self.usage_trim_older_than_days <= 0:
+                self.log.info('Skipping usage trim because usage_trim_older_than_days is not > 0')
+                self.event.sleep(self.usage_trim_interval)
+                continue
+            startDate = '1970-01-01'
+            endDate = (datetime.today() - timedelta(days=self.usage_trim_older_than_days)).strftime('%Y-%m-%d')
+            realms = cast(str, self.usage_trim_realms, default='').split(',')
+            for realm in realms:
+                self.log.info('Running usage trim for realm {} with startDate={} and endDate={}'.format(
+                              realm, startDate, endDate))
```
Comment (rkachach) [inline src/pybind/mgr/rgw/module.py:208]: Usage of f-strings is prefered for formatting: 

`f'Running usage trim for realm {realm} with startDate={startDate} and endDate={endDate}'`

File: doc/mgr/rgw.rst [hunk 2]
```diff
@@ -75,3 +75,32 @@ Update the realm configuration to match the orchestrator deployment
...
+----------------------------
+
+Enable automatic usage log trimming for one or more RGW realms by
+setting the config option ``usage_trim_realms`` to a comma separated
+list of realms.
+
+Set the ``usage_trim_older_than_days`` config option to the amount of
+days usage logs older than this should be trimmed. Optionally you can
+set the interval that trim will be performed with ``usage_trim_interval``,
+it defaults to every twelve hours.
```
Comment (rkachach) [inline doc/mgr/rgw.rst:89]: I think we should specify the units of this value (seems to be seconds! do we really need that fine granularity?)

File: src/pybind/mgr/rgw/module.py [hunk 3]
```diff
@@ -42,7 +43,17 @@ def list_daemons(self, service_name, daemon_type = None, daemon_id = None, host
...
+    MODULE_OPTIONS = [
+        Option(name='usage_trim_realms',
+               default='',
+               desc='Comma separated list with realms to usage trim'),
+        Option(name='usage_trim_older_than_days',
+               default=0,
+               desc='Trim usage log entries older than this amount of days, must be set to a positive value'),
+        Option(name='usage_trim_interval',
+               default=43200,
+               desc='Interval between trimming usage log entries, defaults to 12 hours'),
```
Comment (rkachach) [inline src/pybind/mgr/rgw/module.py:55]: We need to specify the unit of this configuration parameter. Do we really need seconds? isn't hours enough in this case?

File: src/pybind/mgr/rgw/module.py [hunk 4]
```diff
@@ -181,6 +191,27 @@ def _cmd_rgw_realm_reconcile(self,
...
         return HandleCommandResult(retval=retval, stdout=out, stderr=err)
 
+    def serve(self) -> None:
+        self.run = True
+
+        self.log.info('Starting usage trim loop')
+        while self.run:
+            if self.usage_trim_older_than_days <= 0:
+                self.log.info('Skipping usage trim because usage_trim_older_than_days is not > 0')
+                self.event.sleep(self.usage_trim_interval)
```
Comment (rkachach) [inline src/pybind/mgr/rgw/module.py:201]: What happens if somebody configure this parameter to a very big value? the thread could be sleeping here for a very big amount of time and if the user puts a smaller value the new configuration parameter won't be applied!

File: src/pybind/mgr/rgw/module.py (hunk 1 above)
Comment (tobias-urdin) [reply src/pybind/mgr/rgw/module.py:208]: I will fix this!

File: doc/mgr/rgw.rst (hunk 2 above)
Comment (tobias-urdin) [reply doc/mgr/rgw.rst:89]: I will fix this! I think it's kind of hard to anticipate what people want to use it for, could potentially change to minutes but not hours, what do you think?

File: src/pybind/mgr/rgw/module.py (hunk 3 above)
Comment (tobias-urdin) [reply src/pybind/mgr/rgw/module.py:55]: I will fix this!

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 5]
```diff
@@ -367,6 +367,24 @@ def rm_key(self, zone: EntityKey, zg: EntityKey, access_key=None):
...
+class UsageOp:
+    def __init__(self, env):
+        self.env = env
+
+    def trim(self, realm: EntityKey = None, startDate: str = '', endDate: str = ''):
+        ze = ZoneEnv(self.env, realm=realm)
+
+        params = ['trim']
+
+        if startDate != '':
```
[the remark this answers is not in the document]
Comment (tobias-urdin) [reply src/python-common/ceph/rgw/rgwam_core.py:379]: I will fix this so it defaults to None and uses typing to make the str optional.

File: src/pybind/mgr/rgw/module.py (hunk 4 above)
Comment (tobias-urdin) [reply src/pybind/mgr/rgw/module.py:201]: Good feedback! What do we think about a good backoff sleep for when it's not configured? is one minute to small?

File: doc/mgr/rgw.rst (hunk 2 above)
Comment (rkachach) [reply doc/mgr/rgw.rst:89]: The option `usage_trim_older_than_days` uses days as units so I don't see what the benefit of proving more granularity for the option `usage_trim_interval`. I think hours should be OK unless you have some use cases in mind when this wouldn't be enough.

File: src/pybind/mgr/rgw/module.py (hunk 4 above)
Comment (rkachach) [reply src/pybind/mgr/rgw/module.py:201]: In terms of sleep I think 1 min should be OK, probably we have to document that so the user is aware that new values of configuration may take some time to take effect.

File: doc/mgr/rgw.rst (hunk 2 above)
Comment (tobias-urdin) [reply doc/mgr/rgw.rst]: I've made it more clear that it's seconds for the interval and I think the default (12 hours in seconds) is a sensible default that nobody probably wants (needs?) to change but we provide it for convenience either way.

**Pull request closed without merging**